File: apps/api-spring/src/main/java/com/jiandou/api/task/TaskQueueCoordinator.java
package com.jiandou.api.task;

import com.jiandou.api.task.application.port.TaskQueuePort;
import java.util.List;
import org.springframework.stereotype.Component;

@Component
public class TaskQueueCoordinator implements TaskQueuePort {

    private static final int SNAPSHOT_LIMIT = 500;

    private final TaskRepository taskRepository;

    public TaskQueueCoordinator(TaskRepository taskRepository) {
        this.taskRepository = taskRepository;
    }

    @Override
    public void enqueue(String taskId) {
        // Queue state is derived from persisted attempts.
    }

    @Override
    public void remove(String taskId) {
        taskRepository.removeQueuedTask(taskId);
    }

    @Override
    public String claimNext(String workerInstanceId) {
        return taskRepository.claimNextQueuedTask(workerInstanceId);
    }

    @Override
    public List<String> snapshot() {
        return taskRepository.listQueuedTaskIds(SNAPSHOT_LIMIT);
    }
}

File: backend/services/task_queue_coordinator.py
"""Worker-side operations for the persisted task queue."""

from __future__ import annotations

from typing import Any


class TaskQueueCoordinator:
    """Claims, removes, and inspects persisted queue entries."""

    SNAPSHOT_LIMIT = 500

    def __init__(self, task_repository: Any) -> None:
        self._task_repository = task_repository

    async def remove(self, task_id: str) -> None:
        await self._task_repository.remove_queued_task(task_id)

    async def claim_next(self, worker_instance_id: str) -> str | None:
        return await self._task_repository.claim_next_queued_task(worker_instance_id)

    async def snapshot(self) -> list[str]:
        return await self._task_repository.list_queued_task_ids(self.SNAPSHOT_LIMIT)
